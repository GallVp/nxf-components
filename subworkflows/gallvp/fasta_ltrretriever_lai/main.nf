include { SEQKIT_SEQ as UNMASK_IF_ANY   } from '../../../modules/gallvp/seqkit/seq/main'
include { CUSTOM_SHORTENFASTAIDS        } from '../../../modules/gallvp/custom/shortenfastaids/main'
include { LTRHARVEST                    } from '../../../modules/gallvp/ltrharvest/main'
include { LTRFINDER                     } from '../../../modules/gallvp/ltrfinder/main'
include { LTRRETRIEVER_LTRRETRIEVER     } from '../../../modules/gallvp/ltrretriever/ltrretriever/main'
include { CAT_CAT                       } from '../../../modules/gallvp/cat/cat/main'
include { LTRRETRIEVER_LAI              } from '../../../modules/gallvp/ltrretriever/lai/main'
include { CUSTOM_RESTOREGFFIDS          } from '../../../modules/gallvp/custom/restoregffids/main'

workflow FASTA_LTRRETRIEVER_LAI {

    take:
    ch_fasta                        // channel: [ val(meta), fasta ]
    ch_monoploid_seqs               // channel: [ val(meta2), txt ]; Optional: Set to [] if not needed
                                    // val(meta) from ch_fasta and val(meta2) from ch_monoploid_seqs are
                                    // only required to have the same `id`
    skip_lai                        // val(true|false)

    main:
    ch_versions                     = Channel.empty()

    // Prapre input channels
    ch_monoploid_seqs_plain         = ( ch_monoploid_seqs ?: Channel.empty() )
                                    | filter { meta2, seqs -> seqs }
                                    // Cater to channel: [ meta2, [] ]
                                    | map { meta2, seqs -> [ meta2.id, seqs ] }


    // MODULE: SEQKIT_SEQ as UNMASK_IF_ANY
    UNMASK_IF_ANY ( ch_fasta )

    ch_unmasked_fasta               = UNMASK_IF_ANY.out.fastx
    ch_versions                     = ch_versions.mix(UNMASK_IF_ANY.out.versions.first())

    // MOUDLE: CUSTOM_SHORTENFASTAIDS
    CUSTOM_SHORTENFASTAIDS ( ch_unmasked_fasta )

    ch_short_ids_tsv                = CUSTOM_SHORTENFASTAIDS.out.short_ids_tsv
    ch_shortenfastaids_branch       = ch_short_ids_tsv
                                    | branch { meta, tsv ->
                                        change: ! tsv.text.contains('IDs have acceptable length and character')
                                        nonchange: tsv.text.contains('IDs have acceptable length and character')
                                    }

    ch_short_ids_fasta              = ch_shortenfastaids_branch.nonchange
                                    | join(
                                        ch_unmasked_fasta
                                    )
                                    | map { meta, tsv, fasta -> [ meta, fasta ] }
                                    | mix(
                                        ch_shortenfastaids_branch.change
                                        | join(
                                            CUSTOM_SHORTENFASTAIDS.out.short_ids_fasta
                                        )
                                        | map { meta, tsv, fasta -> [ meta, fasta ] }
                                    )

    ch_versions                     = ch_versions.mix(CUSTOM_SHORTENFASTAIDS.out.versions.first())

    // collectFile: Map monoploid seqs to short IDs
    ch_short_monoploid_seqs         = ch_short_ids_tsv
                                    | map { meta, tsv -> [ meta.id, tsv ] }
                                    | join(ch_monoploid_seqs_plain)
                                    | map { id, tsv, seqs ->
                                        map_monoploid_seqs_to_new_ids(id, tsv, seqs)
                                    }
                                    | collectFile(newLine:true)
                                    | map { seqs ->
                                        def id = seqs.name.split('.mapped.monoploid.seqs.txt')[0]

                                        [ id, seqs ]
                                    }
                                    | join(
                                        ch_short_ids_tsv
                                        | map { meta, tsv -> [ meta.id, meta, tsv ] }
                                    )
                                    | map { id, seqs, meta, tsv -> [ meta, seqs ] }


    // MODULE: LTRHARVEST
    LTRHARVEST ( ch_short_ids_fasta )

    ch_ltrharvest_scn               = LTRHARVEST.out.scn
    ch_versions                     = ch_versions.mix(LTRHARVEST.out.versions.first())

    // MODULE: LTRFINDER
    LTRFINDER ( ch_short_ids_fasta )

    ch_ltrfinder_scn                = LTRFINDER.out.scn
    ch_versions                     = ch_versions.mix(LTRFINDER.out.versions.first())

    // MODULE: CAT_CAT
    ch_cat_cat_inputs               = ch_ltrharvest_scn
                                    | join(ch_ltrfinder_scn)
                                    | map { meta, harvested, found -> [ meta, [ harvested, found ] ] }

    CAT_CAT ( ch_cat_cat_inputs )

    ch_ltr_candidates               = CAT_CAT.out.file_out
    ch_versions                     = ch_versions.mix(CAT_CAT.out.versions.first())

    // MODULE: LTRRETRIEVER_LTRRETRIEVER
    ch_ltrretriever_inputs          = ch_short_ids_fasta.join(ch_ltr_candidates)

    LTRRETRIEVER_LTRRETRIEVER (
        ch_ltrretriever_inputs.map { meta, fasta, ltr -> [ meta, fasta ] },
        ch_ltrretriever_inputs.map { meta, fasta, ltr -> ltr },
        [],
        [],
        []
    )

    ch_ltrretriever_log             = LTRRETRIEVER_LTRRETRIEVER.out.log
    ch_pass_list                    = LTRRETRIEVER_LTRRETRIEVER.out.pass_list
    ch_annotation_out               = LTRRETRIEVER_LTRRETRIEVER.out.annotation_out
    ch_pass_out                     = ch_pass_list.join(ch_annotation_out)
    ch_annotation_gff               = LTRRETRIEVER_LTRRETRIEVER.out.annotation_gff
    ch_ltrlib                       = LTRRETRIEVER_LTRRETRIEVER.out.ltrlib
    ch_versions                     = ch_versions.mix(LTRRETRIEVER_LTRRETRIEVER.out.versions.first())

    // MODULE: LTRRETRIEVER_LAI
    ch_short_ids_fasta_mono         = ch_short_ids_fasta
                                    | join(
                                        ch_short_monoploid_seqs,
                                        by:0,
                                        remainder: true
                                    )
                                    // Danger! This partial join can fail
                                    | filter { meta, fasta, seqs -> fasta }
                                    // This filter safeguards against fail on upstream
                                    // process failure: https://github.com/nextflow-io/nextflow/issues/5043
                                    // fasta may come from upstream processes
                                    // seqs also comes from upstream processes, it is optional
                                    // and may not be present for some of the combinations
                                    | map { meta, fasta, seqs -> [ meta, fasta, seqs ?: [] ] }

    ch_lai_inputs                   = skip_lai
                                    ? Channel.empty()
                                    : ch_short_ids_fasta_mono
                                    | join(
                                        ch_pass_out
                                    )
                                    | map { meta, fasta, seqs, pass, out ->
                                        [ meta, fasta, pass, out, seqs ]
                                    }
    LTRRETRIEVER_LAI(
        ch_lai_inputs.map { meta, fasta, pass, out, seqs -> [ meta, fasta ] },
        ch_lai_inputs.map { meta, fasta, pass, out, seqs -> pass },
        ch_lai_inputs.map { meta, fasta, pass, out, seqs -> out },
        ch_lai_inputs.map { meta, fasta, pass, out, seqs -> seqs }
    )

    ch_lai_log                      = LTRRETRIEVER_LAI.out.log
    ch_lai_out                      = LTRRETRIEVER_LAI.out.lai_out
    ch_versions                     = ch_versions.mix(LTRRETRIEVER_LAI.out.versions.first())

    // MODULE: CUSTOM_RESTOREGFFIDS
    ch_gff_tsv_branch               = ch_annotation_gff.join(ch_short_ids_tsv)
                                    | branch { meta, gff, tsv ->
                                        change: ! tsv.text.contains('IDs have acceptable length and character')
                                        nochange: tsv.text.contains('IDs have acceptable length and character')
                                    }

    CUSTOM_RESTOREGFFIDS (
        ch_gff_tsv_branch.change.map { meta, gff, tsv -> [ meta, gff ] },
        ch_gff_tsv_branch.change.map { meta, gff, tsv -> tsv }
    )

    ch_restored_gff                 = ch_gff_tsv_branch.nochange
                                    | map { meta, gff, tsv -> [ meta, gff ] }
                                    | mix(CUSTOM_RESTOREGFFIDS.out.restored_ids_gff3)

    ch_versions                     = ch_versions.mix(CUSTOM_RESTOREGFFIDS.out.versions.first())

    emit:
    ltrretriever_log                = ch_ltrretriever_log   // channel: [ val(meta), fasta ]
    ltrlib                          = ch_ltrlib             // channel: [ val(meta), fasta ]
    annotation_gff                  = ch_restored_gff       // channel: [ val(meta), gff ]
    lai_log                         = ch_lai_log            // channel: [ val(meta), log ]
    lai_out                         = ch_lai_out            // channel: [ val(meta), out ]
    versions                        = ch_versions           // channel: [ versions.yml ]
}


def map_monoploid_seqs_to_new_ids(id, short_ids_tsv, monoploid_seqs) {

    def short_ids_head = short_ids_tsv.text.tokenize('\n')[0]

    if (short_ids_head == "IDs have acceptable length and character. No change required.") {
        return [ "${id}.mapped.monoploid.seqs.txt" ] + monoploid_seqs.text.tokenize('\n')
    }

    def orig_to_new_ids = [:]
    short_ids_tsv.text.eachLine { line ->
        def (original_id, renamed_id) = line.tokenize('\t')
        orig_to_new_ids[original_id] = renamed_id
    }

    def mapped_ids = []
    monoploid_seqs.text.eachLine { original_id ->
        if (!orig_to_new_ids[original_id]) {
            error "Faild to find $original_id in ${short_ids_tsv}" +
            "\nThe short_ids_tsv file is malformed!"
        }

        mapped_ids.add(orig_to_new_ids[original_id])
    }

    return [ "${id}.mapped.monoploid.seqs.txt" ] + mapped_ids
}
