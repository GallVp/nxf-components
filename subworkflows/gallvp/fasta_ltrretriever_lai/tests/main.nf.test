nextflow_workflow {

    name "Test Workflow FASTA_LTRRETRIEVER_LAI"
    script "../main.nf"
    workflow "FASTA_LTRRETRIEVER_LAI"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/fasta_ltrretriever_lai"
    tag "fasta_ltrretriever_lai"
    tag "modules/nf-core/gunzip"
    tag "custom/shortenfastaids"
    tag "ltrharvest"
    tag "ltrfinder"
    tag "ltrretriever/ltrretriever"
    tag "cat/cat"
    tag "ltrretriever/lai"
    tag "custom/restoregffids"

    setup {
        run("GUNZIP") {
            script "../../../../modules/nf-core/gunzip"

            process {
                """
                input[0] = [
                    [ id:'test' ],
                    file(params.modules_testdata_base_path + 'genomics/eukaryotes/actinidia_chinensis/genome/chr1/genome.fasta.gz', checkIfExists: true)
                ]
                """
            }
        }
    }

    test("actinidia_chinensis-genome_1_fasta_gz") {

        when {
            workflow {
                """
                input[0]            = GUNZIP.out.gunzip
                input[1]            = []
                input[2]            = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out.annotation_gff[0][1]).text.contains('Copia_LTR_retrotransposon') },
                { assert file(workflow.out.lai_log[0][1]).text.contains('Calculate LAI:') },
                { assert file(workflow.out.lai_log[0][1]).text.contains('Done!') },
                { assert Math.abs(Float.parseFloat(path(workflow.out.lai_out[0][1]).text.split("\n")[1].split("\t")[6]) - 31.29) <= 1.0 },
                { assert file(workflow.out.ltrlib[0][1]).text.contains('#LTR/Copia') },
                { assert snapshot(workflow.out.versions).match() }
            )
        }
    }

    test("actinidia_chinensis-genome_1_fasta_gz-with_mono") {

        when {
            workflow {
                """
                GUNZIP.out.gunzip.map { meta, fasta ->
                    def fa          = new File('test.fa')
                    fa.write(fasta.text.replaceAll('>chr1', '>chr_xxxxxxxxxxxxxxx_1'))

                    [ meta, fa.toPath() ]
                }
                | set { ch_fa }

                def monoploid_seqs  = new File('test.mono.seq.txt')
                monoploid_seqs.write("chr_xxxxxxxxxxxxxxx_1")

                input[0]            = ch_fa
                input[1]            = Channel.of( [ [ id:'test' ], monoploid_seqs.toPath() ] )
                input[2]            = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out.annotation_gff[0][1]).text.contains('Copia_LTR_retrotransposon') },
                { assert file(workflow.out.lai_log[0][1]).text.contains('Calculate LAI:') },
                { assert file(workflow.out.lai_log[0][1]).text.contains('Done!') },
                { assert Math.abs(Float.parseFloat(path(workflow.out.lai_out[0][1]).text.split("\n")[1].split("\t")[6]) - 31.29) <= 1.0 },
                { assert file(workflow.out.ltrlib[0][1]).text.contains('#LTR/Copia') },
                { assert snapshot(workflow.out.versions).match() }
            )
        }
    }

    test("actinidia_chinensis-genome_1_fasta_gz-with_empty_mono") {

        when {
            workflow {
                """
                GUNZIP.out.gunzip.map { meta, fasta ->
                    def fa          = new File('test.fa')
                    fa.write(fasta.text.replaceAll('>chr1', '>chr_xxxxxxxxxxxxxxx_1'))

                    [ meta, fa.toPath() ]
                }
                | set { ch_fa }

                input[0]            = ch_fa
                input[1]            = []
                input[2]            = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out.annotation_gff[0][1]).text.contains('Copia_LTR_retrotransposon') },
                { assert file(workflow.out.lai_log[0][1]).text.contains('Calculate LAI:') },
                { assert file(workflow.out.lai_log[0][1]).text.contains('Done!') },
                { assert Math.abs(Float.parseFloat(path(workflow.out.lai_out[0][1]).text.split("\n")[1].split("\t")[6]) - 31.29) <= 1.0 },
                { assert file(workflow.out.ltrlib[0][1]).text.contains('#LTR/Copia') },
                { assert snapshot(workflow.out.versions).match() }
            )
        }
    }

    test("empty_fasta_stub") {

        options '-stub'

        when {
            workflow {
                """
                def monoploid_seqs  = new File('test.mono.seq.txt')
                monoploid_seqs.write("chr1")

                input[0]            = GUNZIP.out.gunzip
                input[1]            = Channel.of( [ [ id:'test' ], monoploid_seqs.toPath() ] )
                input[2]            = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }

    test("homo_sapiens - sarscov2 - noltr") {

        when {
            workflow {
                """
                input[0]            = Channel.of([
                    [ id: 'test' ],
                    file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                ])
                input[1]            = []
                input[2]            = false
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { path(workflow.out.ltrretriever_log[0][1]).text.contains('ERROR: No candidate is found in the file(s) you specified.') },
                { assert snapshot(
                    workflow.out.annotation_gff,
                    workflow.out.lai_log,
                    workflow.out.lai_out,
                    workflow.out.ltrlib,
                    workflow.out.versions
                    ).match()
                }
            )
        }
    }
}