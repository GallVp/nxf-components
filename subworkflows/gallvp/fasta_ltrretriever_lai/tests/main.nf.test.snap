{
    "actinidia_chinensis-genome_1_fasta_gz-with_empty_mono": {
        "content": [
            [
                "versions.yml:md5,05a74dacbcd6c5906faeca3206f6563f",
                "versions.yml:md5,4b491828a96bf34cfac8268152ac7a77",
                "versions.yml:md5,52a98b45c3cf73f9314dbe9b34eca9d8",
                "versions.yml:md5,5d86cda262f5c21e43738433a2781ebc",
                "versions.yml:md5,608d746d45680f025562c4be455db461",
                "versions.yml:md5,8468404805dddec679524e75b4fa51e8",
                "versions.yml:md5,9b2d5d76add82441b1e950d333e31307",
                "versions.yml:md5,fb42c6cc5c28b6f9d3384171c7b6b143"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T09:23:40.519096"
    },
    "sarscov2 - fasta - noltr": {
        "content": [
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                
            ],
            [
                "versions.yml:md5,05a74dacbcd6c5906faeca3206f6563f",
                "versions.yml:md5,4b491828a96bf34cfac8268152ac7a77",
                "versions.yml:md5,52a98b45c3cf73f9314dbe9b34eca9d8",
                "versions.yml:md5,608d746d45680f025562c4be455db461",
                "versions.yml:md5,8468404805dddec679524e75b4fa51e8",
                "versions.yml:md5,fb42c6cc5c28b6f9d3384171c7b6b143"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T09:24:10.73061"
    },
    "homo_sapiens - fasta": {
        "content": [
            [
                "versions.yml:md5,05a74dacbcd6c5906faeca3206f6563f",
                "versions.yml:md5,4b491828a96bf34cfac8268152ac7a77",
                "versions.yml:md5,52a98b45c3cf73f9314dbe9b34eca9d8",
                "versions.yml:md5,5d86cda262f5c21e43738433a2781ebc",
                "versions.yml:md5,608d746d45680f025562c4be455db461",
                "versions.yml:md5,8468404805dddec679524e75b4fa51e8",
                "versions.yml:md5,9b2d5d76add82441b1e950d333e31307",
                "versions.yml:md5,fb42c6cc5c28b6f9d3384171c7b6b143"
            ],
            [
                
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T09:36:15.230271"
    },
    "actinidia_chinensis-genome_1_fasta_gz": {
        "content": [
            [
                "versions.yml:md5,05a74dacbcd6c5906faeca3206f6563f",
                "versions.yml:md5,4b491828a96bf34cfac8268152ac7a77",
                "versions.yml:md5,52a98b45c3cf73f9314dbe9b34eca9d8",
                "versions.yml:md5,5d86cda262f5c21e43738433a2781ebc",
                "versions.yml:md5,608d746d45680f025562c4be455db461",
                "versions.yml:md5,8468404805dddec679524e75b4fa51e8",
                "versions.yml:md5,fb42c6cc5c28b6f9d3384171c7b6b143"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T09:17:04.823488"
    },
    "actinidia_chinensis-genome_1_fasta_gz-with_mono": {
        "content": [
            [
                "versions.yml:md5,05a74dacbcd6c5906faeca3206f6563f",
                "versions.yml:md5,4b491828a96bf34cfac8268152ac7a77",
                "versions.yml:md5,52a98b45c3cf73f9314dbe9b34eca9d8",
                "versions.yml:md5,5d86cda262f5c21e43738433a2781ebc",
                "versions.yml:md5,608d746d45680f025562c4be455db461",
                "versions.yml:md5,8468404805dddec679524e75b4fa51e8",
                "versions.yml:md5,9b2d5d76add82441b1e950d333e31307",
                "versions.yml:md5,fb42c6cc5c28b6f9d3384171c7b6b143"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T09:20:22.223941"
    },
    "empty_fasta_stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.LTRlib.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test.out.gff3:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test.LAI.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    [
                        {
                            "id": "test"
                        },
                        "test.LAI.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "5": [
                    "versions.yml:md5,05a74dacbcd6c5906faeca3206f6563f",
                    "versions.yml:md5,4b491828a96bf34cfac8268152ac7a77",
                    "versions.yml:md5,52a98b45c3cf73f9314dbe9b34eca9d8",
                    "versions.yml:md5,5d86cda262f5c21e43738433a2781ebc",
                    "versions.yml:md5,608d746d45680f025562c4be455db461",
                    "versions.yml:md5,8468404805dddec679524e75b4fa51e8",
                    "versions.yml:md5,fb42c6cc5c28b6f9d3384171c7b6b143"
                ],
                "annotation_gff": [
                    [
                        {
                            "id": "test"
                        },
                        "test.out.gff3:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "lai_log": [
                    [
                        {
                            "id": "test"
                        },
                        "test.LAI.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "lai_out": [
                    [
                        {
                            "id": "test"
                        },
                        "test.LAI.out:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "ltrlib": [
                    [
                        {
                            "id": "test"
                        },
                        "test.LTRlib.fa:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "ltrretriever_log": [
                    [
                        {
                            "id": "test"
                        },
                        "test.log:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,05a74dacbcd6c5906faeca3206f6563f",
                    "versions.yml:md5,4b491828a96bf34cfac8268152ac7a77",
                    "versions.yml:md5,52a98b45c3cf73f9314dbe9b34eca9d8",
                    "versions.yml:md5,5d86cda262f5c21e43738433a2781ebc",
                    "versions.yml:md5,608d746d45680f025562c4be455db461",
                    "versions.yml:md5,8468404805dddec679524e75b4fa51e8",
                    "versions.yml:md5,fb42c6cc5c28b6f9d3384171c7b6b143"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "24.04.3"
        },
        "timestamp": "2024-07-29T09:23:53.423404"
    }
}